hibernate-core/src/main/java/org/hibernate/hql/internal/ast/tree/DotNode.java

```diff
@@ -583,39 +577,44 @@ public FromElement getImpliedJoin() {
+	 *
+	 * This method resolves the {@code propertyName} as an property of the entity type at the
+	 * {@link #propertyPath} relative to the {@link #getFromElement() FromElement}.
+	 * The implementation does so by invoking {@link EntityPersister#getPropertyType(String)},
+	 * which will resolve the property path against the entitiy's {@link org.hibernate.persister.entity.PropertyMapping}.
+	 * On initialisation of the {@link EntityPersister}, this {@code PropertyMapping} is filled with
+	 * property paths for all the owned properties and associations, and (embedded) identifier or unique key
+	 * properties for owned associations.
+	 * Henceforth, whenever a property path is found in the {@code PropertyMapping} of the {@code EntityPersister}
+	 * of the {@code FromElement}, we know that the property corresponds to a SQL fragment producible from the
+	 * {@code FromElement}, and as such the entity property can be dereferenced (optimised) in the final query.
+	 *
 	 * <p/>
 	 * For example, consider a fragment like order.customer.id
 	 * (where order is a from-element alias).  Here, we'd have:
 	 * propertyName = "id" AND
-	 * owningType = ManyToOneType(Customer)
-	 * and are being asked to determine whether "customer.id" is a reference
-	 * to customer's PK...
+	 * propertyPath = "customer"
+	 * FromElement = Order
+	 * and are being asked to determine whether "customer.id" is a property path of Order
 	 *
 	 * @param propertyName The name of the property to check.
-	 * @param owningType The type represeting the entity "owning" the property
 	 *
 	 * @return True if propertyName references the entity's (owningType->associatedEntity)
-	 *         primary key; false otherwise.
+	 *         join key; false otherwise.
 	 */
-	private boolean isReferenceToPrimaryKey(String propertyName, EntityType owningType) {
-		EntityPersister persister = getSessionFactoryHelper()
-				.getFactory()
-				.getEntityPersister( owningType.getAssociatedEntityName() );
-		if ( persister.getEntityMetamodel().hasNonIdentifierPropertyNamedId() ) {
-			// only the identifier property field name can be a reference to the associated entity's PK...
-			return propertyName.equals( persister.getIdentifierPropertyName() ) && owningType.isReferenceToPrimaryKey();
+	private boolean isPropertyEmbeddedInJoinProperties(String propertyName) {
+		EntityPersister entityPersister = this.getFromElement().getEntityPersister();
+		String propertyPath = String.join( ".", this.propertyPath, propertyName );
+		try {
+			Type propertyType = entityPersister.getPropertyType( propertyPath );
+			return propertyType != null;
 		}
-		// here, we have two possibilities:
-		// 1) the property-name matches the explicitly identifier property name
-		// 2) the property-name matches the implicit 'id' property name
-		// the referenced node text is the special 'id'
-		if ( EntityPersister.ENTITY_ID.equals( propertyName ) ) {
-			return owningType.isReferenceToPrimaryKey();
+		catch (QueryException e) {
```

Comment:
I supposed you haven't found a way to do this without relying on exception logic? It's something we try to avoid.

Reply:
This cannot be done very easily.

`AbstractEntityPersister` implements `Queryable` which extends `PropertyMapping`. The `AbstractEntityPersister` comprises a `BasicEntityPropertyMapping` in the `propertyMapping` field. Some methods from `PropertyMapping` on the `AbstractEntityPersister`delegate directly to this `BasicEntityPropertyMapping`. So by calling `EntityPersister.getPropertyType(String)`, `BasicEntityPropertyMapping.getPropertyType(String)` is invoked under the hood. This `BasicEntityPropertyMapping` looks up the property in a map, and throws a `QueryException` if the property is not found.

However, the other methods related to properties are implemented somewhat inconsistently.  For example: `getPropertyNames()` on `EntityPersister` does not delegate to the `BasicEntityPropertyMapping`, but instead returns the property names from the metamodel, which does *not* contain the extended property paths. In fact: `PropertyMapping` does not have a method that returns all possible property names.

So in order to make this more neat, I have to add a `getPropertyNames` to `AbstractPropertyMapping`, as well as add a `PropertyMapping` accessor to the `AbstractEntityPersister` or add another delegate method to the `AbstractEntityPersister`. Furthermore, I would have to cast the `EntityPersister` to an `AbstractEntityPersister`. All these changes probably require breaking changes, plus I think that this is one of the aspects of the persister overhaul in 6.0 anyway.